### PR TITLE
Rough artshow + dealers

### PR DESCRIPTION
**Pull request by Gailbear**
rough artshow with some QUALITY fake data. there's still a bunch of formatting work that isn't done, also no TOC and no carousel yet.

image
image

### CODE REVIEW

File: frontend/src/app/_components/creator-page/creator-page.component.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        return 'Website';
+    }
+  }
+
+  // There's gotta be a better way to do this
+  // this is a messy messy sad typecast nightmare
+  linksToDisplay() {
+    if(this.creator && this.creator.links) {
+      const result : {[_: string]: string | undefined} = Object();
+      for (let k of Object.keys(this.creator.links) as unknown as PreferredLink) {
```
Comment (clahey) [inline frontend/src/app/_components/creator-page/creator-page.component.ts:37]: I think the right way to do this is just to iterate over all of them and build result.   Like, addToResult(this.creator.links.etsy, 'Etsy Shop'); Just 5 lines like that.

And then if you want to know what their preferred one was to sort it to the top, have a switch (this.creator.links.preferred) with 5 sections.

It's a bit wordy, but it's a lot easier to read than this and probably easier to write than this.

File: frontend/src/app/_components/creator-page/creator-page.component.html [hunk 2]
```diff
@@ -0,0 +1,21 @@
+<app-breadcrumb [lastCrumbText]="creator.name"></app-breadcrumb>
+<div class="row creator-page">
+    <div class="col-sm-4 col-6">
+        <div class="row">
```
Comment (clahey) [inline frontend/src/app/_components/creator-page/creator-page.component.html:4]: I wouldn't bother with row and col-12 in here and just set your margins and such here.

File: frontend/src/app/_components/creator-page/creator-page.component.html [hunk 3]
```diff
@@ -0,0 +1,21 @@
+<app-breadcrumb [lastCrumbText]="creator.name"></app-breadcrumb>
+<div class="row creator-page">
+    <div class="col-sm-4 col-6">
```
Comment (clahey) [inline frontend/src/app/_components/creator-page/creator-page.component.html:3]: Why have a col-6 and then a col-12? Why not just make them both col-12? Is there any reason for it not to be full width on mobile?

File: frontend/src/app/_components/creator-list/creator-list.component.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import { Creator } from '@app/_models';
+import { Observable } from 'rxjs';
+
+@Component({
+  selector: 'app-creator-list',
+  templateUrl: './creator-list.component.html',
+  styleUrls: ['./creator-list.component.scss']
+})
+export class CreatorListComponent implements OnInit {
+  @Input() creators!: Creator[];
```
Comment (clahey) [inline frontend/src/app/_components/creator-list/creator-list.component.ts:11]: Just set it to [] instead of declaring it ! Maybe? That way it won't crash if they don't pass in a list, it'll just be empty.

File: frontend/src/app/_components/creator-list/creator-list.component.html [hunk 5]
```diff
@@ -0,0 +1,5 @@
+<div class="creator-list row">
+    <div class="col-md-4 col-12 col-sm-6 col-lg-3" *ngFor="let creator of creators">
```
Comment (clahey) [inline frontend/src/app/_components/creator-list/creator-list.component.html:2]: I would list these in order to make it more readable. Doesn't matter if it's col 12 col-sm-6 col-md-4 col-lg3 or the opposite order.

File: frontend/src/app/_components/callout-box/callout-box.component.scss [hunk 6]
```diff
@@ -0,0 +1,13 @@
+.callout-box {
+    background: rgba(116, 171, 244, 0.4);
+
+    .callout-box-title {
```
Comment (clahey) [inline frontend/src/app/_components/callout-box/callout-box.component.scss:4]: It looks like this doesn't work since the classes are in a different component. Not sure what the right way to do it in angular is.

File: frontend/src/app/artshow/artshow.component.html [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+            <div class="callout-box-body">
+                <p>Schedule: link</p>
+                <p>Zoom: link</p>
+            </div>
+        </app-callout-box>
+    </div>
+    <div class="col-md-8 col-12">
+        <div class="row">
+            <div class="col-12">
+                <p>Select the image above an artist's name to view their Art Show gallery.
```
Comment (clahey) [inline frontend/src/app/artshow/artshow.component.html:27]: Maybe just "click on an artist"? You can click on the picture or the name underneath right?

File: frontend/src/app/_components/creator-card/creator-card.component.html [hunk 8]
```diff
@@ -0,0 +1,10 @@
+<div  class="creator card">
+    <img [routerLink]="'/' + component + '/' + creator.id" src="{{creator.images[0].url}}" class="card-img-top" alt="{{creator.images[0].alt}}" />
```
Comment (clahey) [inline frontend/src/app/_components/creator-card/creator-card.component.html:2]: Why not make it a div with the routerLink and make the img and card-body be children of that div?

Comment (clahey) [reply frontend/src/app/_components/creator-card/creator-card.component.html:2]: Oh, or even better, make it an <a> anchor with the routerLink and display:block in the css so that it'll set the cursor correctly?

Comment (clahey) [inline frontend/src/app/_components/creator-card/creator-card.component.html:2]: Would routerLink="/{{component}}/{{creator.id}}" be more readable (and would it work?)

Comment (clahey) [reply frontend/src/app/_components/creator-card/creator-card.component.html:2]: Actually, at the same time, wouldn't [src]="creator.images[0].url" be more readable (and similarly for alt?)

**Review by clahey — APPROVED** [review]

File: frontend/src/app/_components/breadcrumb/breadcrumb.component.ts [hunk 9]
```diff
@@ -9,6 +9,7 @@ import { Router } from '@angular/router';
 export class BreadcrumbComponent implements OnInit {
   crumbs!: {path: string, label: string}[];
   @Input() lobby = true;
+  @Input() lastCrumbText!: string;
```
[the remark this answers is not in the document]
Comment (Gailbear) [reply frontend/src/app/_components/breadcrumb/breadcrumb.component.ts:12]: fixed

File: frontend/src/app/_components/breadcrumb/breadcrumb.component.ts [hunk 10]
```diff
@@ -20,6 +21,9 @@ export class BreadcrumbComponent implements OnInit {
       this.crumbs.unshift({path: '/map', label: 'Lobby'})
     }
     console.log("crumbs", this.crumbs)
```
[the remark this answers is not in the document]
Comment (Gailbear) [reply frontend/src/app/_components/breadcrumb/breadcrumb.component.ts:23]: fixed

File: frontend/src/app/_components/callout-box/callout-box.component.scss [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+        text-align: center;
+        font-size: 3em;
```
[the remark this answers is not in the document]
Comment (Gailbear) [reply frontend/src/app/_components/callout-box/callout-box.component.scss:6]: good catch. fixed.

File: frontend/src/app/_components/callout-box/callout-box.component.scss [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+        font-weight: 200;
+    }
+
+    .callout-box-body {
+        font-size: 2em;
```
[the remark this answers is not in the document]
Comment (Gailbear) [reply frontend/src/app/_components/callout-box/callout-box.component.scss:11]: fixed

File: frontend/src/app/_models/creator.ts [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+        facebook?: string;
+        other?: string;
+        preferred: PreferredLink;
+    }
+    summary: string;
+    description?: string;
+    images: CreatorImage[];
+}
+
+// TODO do we need priority/order for images?
```
[the remark this answers is not in the document]
Comment (Gailbear) [reply frontend/src/app/_models/creator.ts:20]: for us, yes. i was trying to figure out if the people who submitted images cared what order they were in, and whether we were going to preserve that in spreadsheet order or if we needed a separate way of encoding that order.

File: frontend/src/app/_models/creator.ts [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+        etsy?: string;
+        youtube?: string;
+        insta?: string;
+        facebook?: string;
+        other?: string;
+        preferred: PreferredLink;
+    }
+    summary: string;
+    description?: string;
+    images: CreatorImage[];
```
[the remark this answers is not in the document]
Comment (Gailbear) [reply frontend/src/app/_models/creator.ts:17]: we are not hosting the monetary interactions; we decided against building or using a pos system. there are links to people's many avenues of ecommerce on the page.

File: frontend/src/app/_services/index.ts [hunk 15]
```diff
@@ -1,2 +1,3 @@
 export * from './account.service';
-export * from './schedule.service';
\ No newline at end of file
+export * from './schedule.service';
+export * from './creator.service';
```
[the remark this answers is not in the document]
Comment (Gailbear) [reply frontend/src/app/_services/index.ts:3]: hahaha ok yeah

File: frontend/src/app/_services/creator.service.ts [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+export class CreatorService {
+
+  constructor( private http: HttpClient) { }
+
+  make_id_filter(ids?: string[]) : OperatorFunction<Creator[], Creator[]> {
+    if (ids) {
+      return pipe(
+        flatMap(x => of(...x)),
+        filter((c: Creator) => ids.includes(c.id)),
+        toArray()
```
[the remark this answers is not in the document]
Comment (Gailbear) [reply frontend/src/app/_services/creator.service.ts:109]: so, having tested it against loading a static file, it seems to work just fine. this is my solution to "we get the data in a giant array into this filter, we're not getting them split up before they hit here.

File: frontend/src/app/_services/creator.service.ts [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+    return of(fake_data).pipe(
+      this.make_id_filter(ids)
+    );
+    /*return this.http.get<Creator[]>(`${environment.backend}/artists`).pipe(
+      this.make_id_filter(ids)
+    );*/
+  }
+
+  get_dealers(ids?: string[]) : Observable<Creator[]>{
+    return this.http.get<Creator[]>(`${environment.backend}/dealers`).pipe(
```
[the remark this answers is not in the document]
Comment (Gailbear) [reply frontend/src/app/_services/creator.service.ts:125]: good idea

File: frontend/src/app/artshow/artist/artist.component.ts [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+export class ArtistComponent implements OnInit {
+  artist$!: Observable<Creator | undefined>;
+
+  constructor(private route: ActivatedRoute,
+    private creatorService: CreatorService) { }
+
+  ngOnInit(): void {
+    this.artist$ = this.route.params.pipe(
+      pluck('id'),
+      flatMap(id => this.creatorService.get_artists([id])),
```
[the remark this answers is not in the document]
Comment (Gailbear) [reply frontend/src/app/artshow/artist/artist.component.ts:22]: it can't, i don't think. though i don't think changing it will hurt?

File: frontend/src/app/artshow/artshow.component.html [hunk 19]
```diff
@@ -0,0 +1,36 @@
+<app-breadcrumb></app-breadcrumb>
+<div class="row">
+    <div class="col-12">
+        <h1>Welcome to the Art Show!</h1>
+    </div>
+    <div class="col-md-4 col-12">
+        <app-callout-box>
+            <div class="callout-box-title">
```
[the remark this answers is not in the document]
Comment (Gailbear) [reply frontend/src/app/artshow/artshow.component.html:8]: yeah i think it's limiting the css to that component, i'm positive there's a fix, just need to look it up.

File: frontend/src/app/_components/creator-page/creator-page.component.ts [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+  templateUrl: './creator-page.component.html',
+  styleUrls: ['./creator-page.component.scss']
+})
+export class CreatorPageComponent implements OnInit {
+  @Input() creator!: Creator;
+
+  constructor() { }
+
+  ngOnInit(): void {
+  }
```
[the remark this answers is not in the document]
Comment (Gailbear) [reply frontend/src/app/_components/creator-page/creator-page.component.ts:15]: probably! i'll add the todo

File: frontend/src/app/_components/creator-page/creator-page.component.ts (hunk 1 above)
Comment (Gailbear) [reply frontend/src/app/_components/creator-page/creator-page.component.ts:37]: noted! v2. that does seem saner.

File: frontend/src/app/_components/creator-page/creator-page.component.html (hunk 3 above)
Comment (Gailbear) [reply frontend/src/app/_components/creator-page/creator-page.component.html:3]: good call.

File: frontend/src/app/_components/creator-page/creator-page.component.html (hunk 2 above)
Comment (Gailbear) [reply frontend/src/app/_components/creator-page/creator-page.component.html:4]: but then i have to think lol.

File: frontend/src/app/artshow/artshow.component.html (hunk 7 above)
Comment (Gailbear) [reply frontend/src/app/artshow/artshow.component.html:27]: Yeah. I'll change it. Sounds better.

File: frontend/src/app/_components/creator-card/creator-card.component.html (hunk 8 above)
Comment (Gailbear) [reply frontend/src/app/_components/creator-card/creator-card.component.html:2]: it made all the text underlined - also it made the bottom block very confusing. will totally take practical solutions to those problems that aren't this though :)

Comment (Gailbear) [reply frontend/src/app/_components/creator-card/creator-card.component.html:2]: routerLink doesn't seem to take {{}} well, i had tried that one and it was not happy.

i'll change everything re the second comment

File: frontend/src/app/_components/creator-list/creator-list.component.html (hunk 5 above)
Comment (Gailbear) [reply frontend/src/app/_components/creator-list/creator-list.component.html:2]: done

File: frontend/src/app/_models/creator.ts [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+// TODO do we need priority/order for images?
+
+export interface CreatorImage {
+    creator_id: string;
```
[the remark this answers is not in the document]
Comment (Gailbear) [reply frontend/src/app/_models/creator.ts:24]: creator_id is a vestige of the import process that i'm not sure i really want to get rid of for human-aided spreadsheet-based data checking? right now there's no reason to display this separately from the creator, so i would just remove it... but this process is fairly manual

File: frontend/src/app/_services/index.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (Gailbear) [reply frontend/src/app/_services/index.ts:3]: i have done this for all the files i have touched since i read this comment. it's now a setting on my editor. we'll fix the rest of them as we go through.

**Comment by Gailbear** [comment]
no more quality fake data. see slack for more quality real-ish data
